### PR TITLE
#1114 Implement Pair class to ohara

### PR DESCRIPTION
**Pull request by saivirtue**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #1114)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [x] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-common/src/main/java/com/island/ohara/common/data/Pair.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+public final class Pair<L, R> implements Serializable {
+
+  // Serialization version
+  private static final long serialVersionUID = 1L;
+
+  private final L left;
+  private final R right;
+
+  private Pair(final L left, final R right) {
+    this.left = left;
```
Comment (chia7712) [inline ohara-common/src/main/java/com/island/ohara/common/data/Pair.java]: Should we avoid null?

File: ohara-common/src/main/java/com/island/ohara/common/data/Pair.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  public static <L, R> Pair<L, R> of(final L left, final R right) {
+    return new Pair<>(left, right);
+  }
+
+  /**
+   * get the left element from this pair
+   *
+   * @return left element
+   */
+  public L getLeft() {
```
Comment (chia7712) [inline ohara-common/src/main/java/com/island/ohara/common/data/Pair.java]: Could we address scala style. left() and right()

File: ohara-common/src/main/java/com/island/ohara/common/data/Pair.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  /**
+   * update the elements of pair. Will create a new immutable pair
+   *
+   * @param left new left element
+   * @param right new right element
+   * @param <NL> left element type
+   * @param <NR> right element type
+   * @return new immutable pair
+   */
+  public <NL, NR> Pair<NL, NR> update(final NL left, final NR right) {
```
Comment (chia7712) [inline ohara-common/src/main/java/com/island/ohara/common/data/Pair.java]: this method is weird...It just create an new pair which is not associated to origin one.

We should have two **copy** method for Pair. For example.

```java
  public <NR> Pair<L, NR> replaceRight(NR newRight) {
    return new Pair<>(left, newRight);
  } 
```

File: ohara-common/src/main/java/com/island/ohara/common/data/Pair.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  private final R right;
+
+  private Pair(final L left, final R right) {
+    this.left = left;
+    this.right = right;
+  }
+
+  /** Create an immutable pair of nulls */
+  @SuppressWarnings("rawtypes")
+  public static final Pair NULL = Pair.of(null, null);
```
Comment (chia7712) [inline ohara-common/src/main/java/com/island/ohara/common/data/Pair.java]: Is it necessary??

File: ohara-common/src/main/java/com/island/ohara/common/data/Pair.java (hunk 2 above)
Comment (saivirtue) [reply ohara-common/src/main/java/com/island/ohara/common/data/Pair.java]: OK

File: ohara-common/src/main/java/com/island/ohara/common/data/Pair.java (hunk 3 above)
Comment (saivirtue) [reply ohara-common/src/main/java/com/island/ohara/common/data/Pair.java]: In this case , I think user can use
```java
pair.update(pair.getLeft(), new_right)
```

File: ohara-common/src/main/java/com/island/ohara/common/data/Pair.java (hunk 1 above)
Comment (saivirtue) [reply ohara-common/src/main/java/com/island/ohara/common/data/Pair.java]: maybe "empty pair" is less useful...I will remove the relative code

File: ohara-common/src/main/java/com/island/ohara/common/data/Pair.java (hunk 3 above)
Comment (chia7712) [reply ohara-common/src/main/java/com/island/ohara/common/data/Pair.java]: It is almost equal to new Pair(pair.getLeft, new_right), right?

We can just remove this helper method if it does not give a bit benefit.

Comment (saivirtue) [reply ohara-common/src/main/java/com/island/ohara/common/data/Pair.java]: Sure. Update an immutable object is a wield operation...
I will remove this method

**Pull request merged**